roaringbitmap/src/main/java/org/roaringbitmap/AdaptiveOrderedWriter.java

```diff
@@ -0,0 +1,59 @@
+package org.roaringbitmap;
+
+public class AdaptiveOrderedWriter implements OrderedWriter {
+  private SparseOrderedWriter sparseWriter;
+  private DenseOrderedWriter denseWriter;
+  private boolean sparse;
+
+  public AdaptiveOrderedWriter(RoaringBitmap underlying) {
+    this.sparseWriter = new SparseOrderedWriter(underlying);
+    this.sparse = true;
+  }
+
+  AdaptiveOrderedWriter(RoaringBitmap underlying, int size) {
+    this.sparseWriter = new SparseOrderedWriter(underlying, size);
+    this.sparse = true;
+  }
+
+  @Override
+  public void add(int value) {
+    if (sparse) {
+      try {
+        sparseWriter.add(value);
+      } catch (IllegalStateException e) {
+        denseWriter = sparseWriter.transfer(denseWriter);
+        denseWriter.add(value);
+        sparse = false;
+      }
+    } else {
+      denseWriter.add(value);
+      if (!denseWriter.isDirty()) {
```

Comment:
I don't see any evidence that switching back to sparse is a good idea, and probably makes the code slower.